pkg/ctrl/doc.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package ctrl provides libraries for building Controllers.  Controllers implement Kubernetes APIs
+and are central to building Operators, Workload APIs, Configuration APIs, Autoscalers, and more.
+
+Controllers
+
+Controllers are work queues that watch for changes to objects (i.e. Create / Update / Delete Events) and
+then make changes to the cluster in response (i.e. Reconcile the object).  Controllers DO NOT react directly
+to an event, but instead enqueue a ReconcileRequest.  When the ReconcileRequest is processed, the Reconcile
+will read a specific object and compare the state requested in that object against the actual cluster state.
+It will then make changes based upon the difference in what is requested from what is observed.
+
+Controller Workflow
+
+	1. Watch one or more Sources of Events (i.e. changes to Kubernetes objects resulting from Create / Update / Delete operations).
+
+	  1.1 EventHandlers enqueue ReconcileRequests (containing the Names and Namespaces of objects to Reconcile).
+
+	2. Dequeue a ReconcileRequest and Reconcile an object.
+
+	  2.1 Read the object referred to in the request as well as any related objects (e.g. owned objects).
+
+	  2.2 Create or update Kubernetes objects as needed to execute the desired state specified in the object.
```

Comment:
... Kubernetes objects and/or external state...

Reply:
Done